src/main/java/modbus/SlaveNode.java

```diff
@@ -298,7 +212,138 @@ public void readPoints() {
+		if (dataType == DataType.BOOLEAN && !BinaryLocator.isBinaryRange(range) && bit < 0) {
+			dt = com.serotonin.modbus4j.code.DataType.TWO_BYTE_INT_SIGNED;
+		}
+
+		BaseLocator<?> locator = BaseLocator.createLocator(slaveId, range, offset, dt, bit, numRegs);
+		return locator;
+	}
+	
+	private void updateValue(Node pnode, Object obj) {
+		DataType dataType = DataType.valueOf(pnode.getAttribute(ATTR_DATA_TYPE).getString());
+		double scaling = Util.getDoubleValue(pnode.getAttribute(ATTR_SCALING));
+		double addscale = Util.getDoubleValue(pnode.getAttribute(ATTR_SCALING_OFFSET));
+
+		ValueType vt = null;
+		Value v = null;
+		if (DataType.getDataTypeInt(dataType) != null) {
+			if (dataType == DataType.BOOLEAN && obj instanceof Boolean) {
+				vt = ValueType.BOOL;
+				v = new Value((Boolean) obj);
+			} else if (dataType == DataType.BOOLEAN && obj instanceof Number) {
+				vt = ValueType.ARRAY;
+				JsonArray jarr = new JsonArray();
+				for (int i = 0; i < 16; i++) {
+					jarr.add(((((Number) obj).intValue() >> i) & 1) == 1);
+				}
+				v = new Value(jarr);
+			} else if (dataType.isString() && obj instanceof String) {
+				vt = ValueType.STRING;
+				v = new Value((String) obj);
+			} else if (obj instanceof Number) {
+				vt = ValueType.NUMBER;
+				Number num = (Number) obj;
+				v = new Value(num.doubleValue() / scaling + addscale);
+			} else if (obj instanceof ExceptionResult) {
+				ExceptionResult result = (ExceptionResult) obj;
+				LOGGER.error(pnode.getName() + " : " + result.getExceptionMessage());
+			}
+		} else {
+			switch (dataType) {
+			case INT32M10KSWAP:
+			case INT32M10K: {
+				short shi = (short) (((Number) obj).intValue() >>> 16);
+				short slo = (short) (((Number) obj).intValue() & 0xffff);
+				boolean swap = (dataType == DataType.INT32M10KSWAP);
+				int num;
+				if (swap)
+					num = ((int) slo) * 10000 + (int) shi;
+				else
+					num = ((int) shi) * 10000 + (int) slo;
+				vt = ValueType.NUMBER;
+				v = new Value(num / scaling + addscale);
+				break;
+			}
+			case UINT32M10KSWAP:
+			case UINT32M10K: {
+				short shi = (short) (((Number) obj).intValue() >>> 16);
+				short slo = (short) (((Number) obj).intValue() & 0xffff);
+				boolean swap = (dataType == DataType.INT32M10KSWAP);
+				long num;
+				if (swap)
+					num = Util.toUnsignedLong(Util.toUnsignedInt(slo) * 10000 + Util.toUnsignedInt(shi));
+				else
+					num = Util.toUnsignedLong(Util.toUnsignedInt(shi) * 10000 + Util.toUnsignedInt(slo));
+				vt = ValueType.NUMBER;
+				v = new Value(num / scaling + addscale);
+				break;
+			}
+			default:
+				vt = null;
+				v = null;
+				break;
+			}
+		}
+
+		if (v != null) {
+			pnode.setValueType(vt);
+			pnode.setValue(v);
+		} else if (node.getAttribute(ModbusConnection.ATTR_ZERO_ON_FAILED_POLL).getBool()) {
+			if (pnode.getValueType().compare(ValueType.NUMBER)) {
+				pnode.setValue(new Value(0));
+			} else if (pnode.getValueType().compare(ValueType.BOOL)) {
+				pnode.setValue(new Value(false));

```

Comment:
there are many magic numbers in here. I'd suggest extracting some constants to make the code easier to understand.